{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutational Signature Extraction\n",
    "In this notebook, we will extract SNV and indel signatures for the metastatic pheochromocytoma and paraganglioma cohort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required packages\n",
    "import os\n",
    "import csv\n",
    "from SigProfilerMatrixGenerator.scripts import SigProfilerMatrixGeneratorFunc as matGen\n",
    "from SigProfilerExtractor import sigpro as sig\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set global variants\n",
    "read_depth = 20\n",
    "genome = 'GrCh38'\n",
    "project = 'mPPGL'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create SigProfilerMatrixGenerator Input\n",
    "We will use basic data wrangling to create the required input for SigProfilerMatrixGenerator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "snvs = [file for file in os.listdir(\"../../data/raw/snvs/\") if \".csv\" in file] # Create list of SNV files\n",
    "samples = [sample.split(\".\")[0] for sample in snvs] # Extract sample names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create input files\n",
    "outfields = ['Project', 'Sample', 'ID', 'Genome', 'mut_type', 'chrom', 'pos_start', 'pos_end', 'ref', 'alt', 'type']\n",
    "\n",
    "for snv, sample in zip(snvs, samples):\n",
    "    with open(\"../../data/raw/snvs/{0}\".format(snv), 'r') as infile, open(\"../../data/processed/mutational_signatures/input/{0}.SigProfiler.txt\".format(sample), 'w') as outfile:\n",
    "    \n",
    "        reader = csv.DictReader(infile, delimiter=',')\n",
    "        writer = csv.DictWriter(outfile, delimiter='\\t', fieldnames=outfields, lineterminator='\\r')\n",
    "        writer.writeheader()\n",
    "\n",
    "        for row in reader:\n",
    "\n",
    "            if float(row['Tumor.AltDepth']) >= float(read_depth) and row['MUTECT2'] == \"PASS\":\n",
    "\n",
    "                newDict = {'Project': project,\n",
    "                        'Sample': row['Tumor.ID'],\n",
    "                        'ID': '.', \n",
    "                        'Genome': genome,\n",
    "                        'mut_type': row['Variant.Class'],\n",
    "                        'chrom': row['Chr'],\n",
    "                        'pos_start': row['Start'],\n",
    "                        'pos_end': row['Start'], \n",
    "                        'ref': row['REF'],\n",
    "                        'alt': row['ALT'],\n",
    "                        'type': 'SOMATIC'}\n",
    "\n",
    "                writer.writerow(newDict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run SigProfilerMatrixGenerator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 248.64 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 44.34 seconds.\n",
      "Matrices generated for 44 samples with 846 errors. Total of 2122 SNVs, 19 DINUCs, and 748 INDELs were successfully analyzed.\n"
     ]
    }
   ],
   "source": [
    "matrices = matGen.SigProfilerMatrixGeneratorFunc(project, \n",
    "                                                 genome, \n",
    "                                                 \"../../data/processed/mutational_signatures/input\",\n",
    "                                                 plot=True, exome=False, bed_file=None, chrom_based=False, \n",
    "                                                 tsb_stat=True, seqInfo=True, cushion=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract SBS Signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MutationType</th>\n",
       "      <th>PP002-DZ3A</th>\n",
       "      <th>PP002-DZ4A</th>\n",
       "      <th>PP002-DZ5A</th>\n",
       "      <th>PP008-DZ2A</th>\n",
       "      <th>PP009-DZ1B</th>\n",
       "      <th>PP009-DZ2B</th>\n",
       "      <th>PP010-DZ1A</th>\n",
       "      <th>PP010-DZ5A</th>\n",
       "      <th>PP010-F2</th>\n",
       "      <th>...</th>\n",
       "      <th>PP253-DZ1A</th>\n",
       "      <th>PP253-DZ2A</th>\n",
       "      <th>PP263-DZ2A</th>\n",
       "      <th>PP302-DZ1A</th>\n",
       "      <th>PP323-DZ3A</th>\n",
       "      <th>PP323-DZ4A</th>\n",
       "      <th>PP323-DZ5A</th>\n",
       "      <th>PP341-DZ2B</th>\n",
       "      <th>PP344-DZ1A</th>\n",
       "      <th>PP344-DZ3A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A[C&gt;A]C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A[C&gt;A]G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A[C&gt;A]T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A[C&gt;G]A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  MutationType  PP002-DZ3A  PP002-DZ4A  PP002-DZ5A  PP008-DZ2A  PP009-DZ1B  \\\n",
       "0      A[C>A]A           0           0           0           3           0   \n",
       "1      A[C>A]C           0           0           0           3           0   \n",
       "2      A[C>A]G           0           0           0           1           0   \n",
       "3      A[C>A]T           0           0           0           1           0   \n",
       "4      A[C>G]A           0           0           0           0           0   \n",
       "\n",
       "   PP009-DZ2B  PP010-DZ1A  PP010-DZ5A  PP010-F2  ...  PP253-DZ1A  PP253-DZ2A  \\\n",
       "0           0           0           0         0  ...           1           0   \n",
       "1           0           0           0         0  ...           0           0   \n",
       "2           0           0           0         0  ...           0           0   \n",
       "3           0           0           0         0  ...           0           0   \n",
       "4           0           0           0         0  ...           0           0   \n",
       "\n",
       "   PP263-DZ2A  PP302-DZ1A  PP323-DZ3A  PP323-DZ4A  PP323-DZ5A  PP341-DZ2B  \\\n",
       "0           0           0           1          17           4           0   \n",
       "1           0           2           0           9           3           0   \n",
       "2           0           0           0           7           2           0   \n",
       "3           0           0           0          12           0           0   \n",
       "4           0           0           0           1           1           0   \n",
       "\n",
       "   PP344-DZ1A  PP344-DZ3A  \n",
       "0           0           1  \n",
       "1           0           2  \n",
       "2           0           0  \n",
       "3           0           0  \n",
       "4           1           0  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data\n",
    "df1_path = \"../../data/processed/mutational_signatures/input/output/SBS/mPPGL.SBS96.all\"\n",
    "df1 = pd.read_table(df1_path)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************** Reported Current Memory Use: 1.17 GB *****************\n",
      "\n",
      "Extracting signature 1 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 1 signatures is 50.67 seconds\n",
      "Optimization time is 2.821291923522949 seconds\n",
      "The reconstruction error is 0.7388, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 1 signatures\n",
      "\n",
      "\n",
      "Extracting signature 2 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 2 signatures is 68.21 seconds\n",
      "Optimization time is 2.6764321327209473 seconds\n",
      "The reconstruction error is 0.4993, average process stability is 0.96 and \n",
      "the minimum process stability is 0.95 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 3 signatures is 123.71 seconds\n",
      "Optimization time is 2.747195243835449 seconds\n",
      "The reconstruction error is 0.2539, average process stability is 0.88 and \n",
      "the minimum process stability is 0.78 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 4 signatures is 243.65 seconds\n",
      "Optimization time is 2.778985023498535 seconds\n",
      "The reconstruction error is 0.1918, average process stability is 0.74 and \n",
      "the minimum process stability is 0.36 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 5 signatures is 2179.78 seconds\n",
      "Optimization time is 6.796194076538086 seconds\n",
      "The reconstruction error is 0.18, average process stability is 0.6 and \n",
      "the minimum process stability is -0.02 for 5 signatures\n",
      "\n",
      "\n",
      "Extracting signature 6 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 6 signatures is 2938.61 seconds\n",
      "Optimization time is 4.875287771224976 seconds\n",
      "The reconstruction error is 0.1753, average process stability is 0.47 and \n",
      "the minimum process stability is -0.18 for 6 signatures\n",
      "\n",
      "\n",
      "Extracting signature 7 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 7 signatures is 558.9 seconds\n",
      "Optimization time is 3.830271005630493 seconds\n",
      "The reconstruction error is 0.1659, average process stability is 0.37 and \n",
      "the minimum process stability is -0.32 for 7 signatures\n",
      "\n",
      "\n",
      "Extracting signature 8 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 8 signatures is 571.11 seconds\n",
      "Optimization time is 3.9189157485961914 seconds\n",
      "The reconstruction error is 0.1603, average process stability is 0.32 and \n",
      "the minimum process stability is -0.33 for 8 signatures\n",
      "\n",
      "\n",
      "Extracting signature 9 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 9 signatures is 685.25 seconds\n",
      "Optimization time is 3.993483066558838 seconds\n",
      "The reconstruction error is 0.1504, average process stability is 0.26 and \n",
      "the minimum process stability is -0.33 for 9 signatures\n",
      "\n",
      "\n",
      "Extracting signature 10 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 10 signatures is 558.08 seconds\n",
      "Optimization time is 3.574387788772583 seconds\n",
      "The reconstruction error is 0.1371, average process stability is 0.24 and \n",
      "the minimum process stability is -0.2 for 10 signatures\n",
      "\n",
      "\n",
      "The selected opportunity genome is GrCh38. COSMIC signatures are available only for GRCh37/38, mm9/10 and rn6 genomes. So, the opportunity genome is reset to GRCh37.\n",
      "\n",
      " Denovo Fitting .....\n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 44/44 [100%] in 0.3s (146.87/s) \n",
      "\n",
      " Decomposing De Novo Signatures  .....\n",
      "Decompositon Plot:SBS96A |â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰| 1/1 [100%] in 0.0s (23967.60/s) \n",
      "Decompositon Plot:SBS96B |â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰| 1/1 [100%] in 0.0s (24850.28/s) \n",
      "Decompositon Plot:SBS96C |â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰| 1/1 [100%] in 0.0s (20929.70/s) \n",
      "\n",
      " Assigning decomposed signature\n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 44/44 [100%] in 0.4s (101.88/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfiler Extractor.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# Run extraction\n",
    "sig.sigProfilerExtractor(\n",
    "        \"matrix\",\n",
    "        \"../../data/processed/mutational_signatures/SigProfilerExtractor_SBS\",\n",
    "        df1,\n",
    "        opportunity_genome=\"GrCh38\",\n",
    "        exome=True,\n",
    "        minimum_signatures=1,\n",
    "        maximum_signatures=10,\n",
    "        cpu=4,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Indel Signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MutationType</th>\n",
       "      <th>PP002-DZ3A</th>\n",
       "      <th>PP002-DZ4A</th>\n",
       "      <th>PP002-DZ5A</th>\n",
       "      <th>PP008-DZ2A</th>\n",
       "      <th>PP009-DZ1B</th>\n",
       "      <th>PP009-DZ2B</th>\n",
       "      <th>PP010-DZ1A</th>\n",
       "      <th>PP010-DZ5A</th>\n",
       "      <th>PP010-F2</th>\n",
       "      <th>...</th>\n",
       "      <th>PP253-DZ1A</th>\n",
       "      <th>PP253-DZ2A</th>\n",
       "      <th>PP263-DZ2A</th>\n",
       "      <th>PP302-DZ1A</th>\n",
       "      <th>PP323-DZ3A</th>\n",
       "      <th>PP323-DZ4A</th>\n",
       "      <th>PP323-DZ5A</th>\n",
       "      <th>PP341-DZ2B</th>\n",
       "      <th>PP344-DZ1A</th>\n",
       "      <th>PP344-DZ3A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:Del:C:0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:Del:C:1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:Del:C:2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:Del:C:3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:Del:C:4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  MutationType  PP002-DZ3A  PP002-DZ4A  PP002-DZ5A  PP008-DZ2A  PP009-DZ1B  \\\n",
       "0    1:Del:C:0           0           0           0           1           0   \n",
       "1    1:Del:C:1           0           0           0           0           0   \n",
       "2    1:Del:C:2           0           0           0           0           0   \n",
       "3    1:Del:C:3           0           0           0           0           0   \n",
       "4    1:Del:C:4           0           0           0           0           0   \n",
       "\n",
       "   PP009-DZ2B  PP010-DZ1A  PP010-DZ5A  PP010-F2  ...  PP253-DZ1A  PP253-DZ2A  \\\n",
       "0           0           0           0         0  ...           0           0   \n",
       "1           0           0           0         0  ...           0           0   \n",
       "2           0           0           0         0  ...           0           0   \n",
       "3           0           0           0         0  ...           0           0   \n",
       "4           0           0           0         0  ...           0           0   \n",
       "\n",
       "   PP263-DZ2A  PP302-DZ1A  PP323-DZ3A  PP323-DZ4A  PP323-DZ5A  PP341-DZ2B  \\\n",
       "0           0           0           0           1           0           0   \n",
       "1           0           0           0           0           0           0   \n",
       "2           0           0           0           0           0           0   \n",
       "3           0           0           0           0           0           0   \n",
       "4           0           0           0           0           0           0   \n",
       "\n",
       "   PP344-DZ1A  PP344-DZ3A  \n",
       "0           0           0  \n",
       "1           0           0  \n",
       "2           0           0  \n",
       "3           0           0  \n",
       "4           0           0  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data\n",
    "df2_path = \"../../data/processed/mutational_signatures/input/output/ID/mPPGL.ID83.all\"\n",
    "df2 = pd.read_table(df2_path)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************** Reported Current Memory Use: 1.53 GB *****************\n",
      "\n",
      "Extracting signature 1 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 1 signatures is 76.49 seconds\n",
      "Optimization time is 4.839684963226318 seconds\n",
      "The reconstruction error is 0.0674, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 1 signatures\n",
      "\n",
      "\n",
      "Extracting signature 2 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 2 signatures is 111.64 seconds\n",
      "Optimization time is 4.449145793914795 seconds\n",
      "The reconstruction error is 0.0535, average process stability is 0.85 and \n",
      "the minimum process stability is 0.69 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 3 signatures is 155.08 seconds\n",
      "Optimization time is 4.411947011947632 seconds\n",
      "The reconstruction error is 0.0688, average process stability is 0.42 and \n",
      "the minimum process stability is -0.21 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 4 signatures is 218.56 seconds\n",
      "Optimization time is 5.79591703414917 seconds\n",
      "The reconstruction error is 0.0797, average process stability is 0.26 and \n",
      "the minimum process stability is -0.6 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 5 signatures is 282.43 seconds\n",
      "Optimization time is 4.0458478927612305 seconds\n",
      "The reconstruction error is 0.0496, average process stability is 0.18 and \n",
      "the minimum process stability is -0.75 for 5 signatures\n",
      "\n",
      "\n",
      "Extracting signature 6 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 6 signatures is 430.77 seconds\n",
      "Optimization time is 5.7666871547698975 seconds\n",
      "The reconstruction error is 0.0352, average process stability is 0.13 and \n",
      "the minimum process stability is -0.85 for 6 signatures\n",
      "\n",
      "\n",
      "Extracting signature 7 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 7 signatures is 473.77 seconds\n",
      "Optimization time is 4.087454795837402 seconds\n",
      "The reconstruction error is 0.0338, average process stability is 0.05 and \n",
      "the minimum process stability is -0.88 for 7 signatures\n",
      "\n",
      "\n",
      "Extracting signature 8 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 8 signatures is 628.02 seconds\n",
      "Optimization time is 4.518688917160034 seconds\n",
      "The reconstruction error is 0.0402, average process stability is 0.02 and \n",
      "the minimum process stability is -0.81 for 8 signatures\n",
      "\n",
      "\n",
      "Extracting signature 9 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 9 signatures is 813.52 seconds\n",
      "Optimization time is 3.9379119873046875 seconds\n",
      "The reconstruction error is 0.0294, average process stability is 0.02 and \n",
      "the minimum process stability is -0.78 for 9 signatures\n",
      "\n",
      "\n",
      "Extracting signature 10 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 10 signatures is 1009.47 seconds\n",
      "Optimization time is 2.9672980308532715 seconds\n",
      "The reconstruction error is 0.052, average process stability is -0.02 and \n",
      "the minimum process stability is -0.79 for 10 signatures\n",
      "\n",
      "\n",
      "The selected opportunity genome is GrCh38. COSMIC signatures are available only for GRCh37/38, mm9/10 and rn6 genomes. So, the opportunity genome is reset to GRCh37.\n",
      "\n",
      " Denovo Fitting .....\n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 27/27 [100%] in 0.2s (166.28/s) \n",
      "\n",
      " Decomposing De Novo Signatures  .....\n",
      "Decompositon Plot:ID83A |â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰| 1/1 [100%] in 0.0s (15376.10/s) \n",
      "Decompositon Plot:ID83B |â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰â–‰| 1/1 [100%] in 0.0s (23230.96/s) \n",
      "\n",
      " Assigning decomposed signature\n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 27/27 [100%] in 0.2s (144.12/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfiler Extractor.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "sig.sigProfilerExtractor(\n",
    "        \"matrix\",\n",
    "        \"../../data/processed/mutational_signatures/SigProfilerExtractor_ID\",\n",
    "        df2,\n",
    "        opportunity_genome=\"GrCh38\",\n",
    "        exome=True,\n",
    "        minimum_signatures=1,\n",
    "        maximum_signatures=10,\n",
    "        cpu=4,\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
